### PR TITLE
tree: Fix Recursive MapNode d.ts generation

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

Workaround issue where recursive tree schemas using MapNodes produce invalid d.ts files.

See changeset for details.

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

### CODE REVIEW

File: experimental/framework/tree-react-api/src/testExports.ts [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+export const recursiveField = schema.optionalRecursive([() => RecursiveObject, schema.number]);
+
+export class RecursiveMap extends schema.mapRecursive("RM", [() => RecursiveMap]) {}
+{
+	type _check = ValidateRecursiveSchema<typeof RecursiveMap>;
+}
+
+// TODO: export once ` error TS2310: Type 'RecursiveArray' recursively references itself as a base type.` in the d.ts file is fixed.
```
Comment (Josmithr) [inline experimental/framework/tree-react-api/src/testExports.ts:98]: Similar issue to the map one being addressed here? Just planned for a subsequent PR, presumably?

File: packages/dds/tree/api-report/tree.beta.api.md [hunk 2]
```diff
@@ -252,6 +253,22 @@ type ObjectFromSchemaRecordUnsafe<T extends Unenforced<RestrictiveReadonlyRecord
 // @public
 export type Off = () => void;
 
+// @public @sealed
+interface ReadonlyMapInlined<K, T extends Unenforced<ImplicitAllowedTypes>> {
```
Comment (Josmithr) [inline packages/dds/tree/api-report/tree.beta.api.md:257]: Should this be a part of the internal types namespace export?

Comment (Josmithr) [reply packages/dds/tree/api-report/tree.beta.api.md:257]: Ah, or was that just replaced with `@system`?

File: packages/dds/tree/src/test/simple-tree/typesUnsafe.spec.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+ * Licensed under the MIT License.
+ */
+
+// eslint-disable-next-line import/no-internal-modules
+import type { ReadonlyMapInlined } from "../../simple-tree/typesUnsafe.js";
+// eslint-disable-next-line import/no-internal-modules
+import type { numberSchema } from "../../simple-tree/leafNodeSchema.js";
+import type { areSafelyAssignable, requireTrue } from "../../util/index.js";
+
+type MapInlined = ReadonlyMapInlined<string, typeof numberSchema>;
```
Comment (Josmithr) [inline packages/dds/tree/src/test/simple-tree/typesUnsafe.spec.ts:12]: Nit: `ReadonlyMapInlined` is defined in `internalTypes` - should this test be in `internalTypes.spec.ts`?

**Review by Josmithr — APPROVED** [review]

File: experimental/framework/tree-react-api/src/testExports.ts (hunk 1 above)
Comment (CraigMacomber) [reply experimental/framework/tree-react-api/src/testExports.ts:98]: I'll have a follow-up PR with a workaround for that case. Sadly the same approach didn't fix it so I did something less nice.

File: packages/dds/tree/api-report/tree.beta.api.md (hunk 2 above)
Comment (CraigMacomber) [reply packages/dds/tree/api-report/tree.beta.api.md:257]: It is part of internalTypes (which is a module not a namespace currently). I redundantly put system and sealed on it for extra clarity.

Comment (CraigMacomber) [reply packages/dds/tree/api-report/tree.beta.api.md:257]: Note in this APi report, this definition is not exported: its exported from the namespace (thats how API extractor handles namespaces)

File: packages/dds/tree/src/test/simple-tree/typesUnsafe.spec.ts (hunk 3 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/test/simple-tree/typesUnsafe.spec.ts:12]: the code this is testing is in the file this is the test for (typesUnsafe.ts)

internalTypes just has a reexport for it, and this is not testing the reexport, its testing the definition.

**Pull request merged**